src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,474 @@
+    refresh_in_progress: Arc<Mutex<()>>,
+
+    // The token provider. This thing does the refreshing.
+    inner: Arc<T>,
+}
+
+// We manually implement the `Clone` trait because the Rust compiler will
+// squawk if `T` is not `Clone`, even though we only hold an `Arc<T>`. :shrug:
+impl<T: TokenProvider> Clone for TokenCache<T> {
+    fn clone(&self) -> TokenCache<T> {
+        TokenCache {
+            token: self.token.clone(),
+            refresh_in_progress: self.refresh_in_progress.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+fn initial_token() -> Result<Token> {
+    Err(CredentialError::retryable_from_str("No token or error in the cache. This should never happen. Something has gone wrong. Open an issue at https://github.com/googleapis/google-cloud-rust/issues/new?template=bug_report.md."))
+}
+
+impl<T: TokenProvider> TokenCache<T> {
+    // TODO(#1210) - implementation in progress
+    #[allow(dead_code)]
+    pub fn new(inner: T) -> TokenCache<T> {
+        TokenCache {
+            token: Arc::new(RwLock::new(initial_token())),
+            refresh_in_progress: Arc::new(Mutex::new(())),
+            inner: Arc::new(inner),
+        }
+    }
+
+    fn state(&self) -> TokenState {
+        let token = self.token.read().unwrap();
+        match &*token {
+            Ok(t) => match t.expires_at {
+                None => TokenState::Valid(t.clone()),
+                Some(e) => {
+                    let now = Instant::now().into_std();
+                    if e <= now {
+                        TokenState::Invalid(self.try_refresh_lock())
+                    } else {
+                        TokenState::Valid(t.clone())
+                    }
+                }
+            },
+            Err(_) => TokenState::Invalid(self.try_refresh_lock()),
+        }
+    }
+
+    fn try_refresh_lock(&self) -> RefreshLock {
+        self.refresh_in_progress.clone().try_lock_owned()
+    }
+}
+
+type RefreshLock = std::result::Result<tokio::sync::OwnedMutexGuard<()>, tokio::sync::TryLockError>;
+
+enum TokenState {
+    Invalid(RefreshLock),
+    Valid(Token),
+}
+
+#[async_trait::async_trait]
+impl<T: TokenProvider + 'static> TokenProvider for TokenCache<T> {
+    async fn get_token(&self) -> Result<Token> {
+        match self.state() {
+            TokenState::Valid(token) => {
+                // We have a valid token, and it is not expiring any time soon.
+                // Return it immediately.
+                Ok(token)
+            }
+
+            TokenState::Invalid(refresh_guard) => {
+                match refresh_guard {
+                    // Check if there are any outstanding refreshes...
+                    Ok(_guard) => {
+                        // No refreshes. We should start one.
+                        let token = self.inner.get_token().await;
+
+                        // Store the token, or an updated error.
+                        *self.token.write().unwrap() = token.clone();
```

Comment:
Now T2 is trying to acquire the write lock on `self.token`.  It cannot because T1 is holding a read lock... 

I think you have a deadlock.

Reply:
The read lock in T1 is released when we return from `state()`. No?

Reply:
`T1` is in the body of `state()` it has not returned.  It is trying to acquire the `refresh_lock`...

Reply:
> It is trying to acquire the refresh_lock...

key word: trying.

Note that we call `try_lock_owned()`, not `lock()`

https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.try_lock_owned

> `T1` is in the body of `state()`

If `T2` is holding the refresh lock is held, we return a `TryLockError` and exit `self.state()`